### ISSUE TITLE
[feature] Expose cmake_layout:build_folder as a setting

### ISSUE THREAD

**Issue by jjcasmar** [conan-io/conan#13829]
### What is your suggestion?

Currently, when using `cmake_layout`, the build folder is defined by using a combination of several variables. First, there is a `build_folder` variable, which just sets a path. There is also the `build_folder_vars` variable, which take into account several variables specified as a setting. Depending on the build generator (Ninjca, Makefile...), is also takes into account the build type or not in the final build folder.

By combining those two methods, you can set the build folder to be `<root>/<cmake_layout:build_folder>/<path sepcified by variables>/<build_type>`, obtaining `<root>/build/desktop/linux-x86_64-gcc-10/Release` for example.

However, this is not enough, as different builds my set the build_folder, for example, doing `-s *:build_type=Debug -s &:build_type=Debug` or `-s *:build_type=Release -s &:build_type=Debug` may define the same build_folder name. Also, if I have several options and I want different `conan install` commands with different options, if is very difficult to track the name of the folder.

I think adding the option to directly set the build folder would simplify this, as we could match a name which makes semantic sense, to a specific build folder, with specific options. For example, I might make sense to define a folder `build/benchmarking` instead of a folder `build/linux-x86_64-clang-16-use_timers_True_is_benchmarking_True/Release`

The issue is that the `cmake_layout:build_folder` is not exposed as a setting, so in order to modify it, you need to directly tweak the conanfile. To do so I added a hack into my conanfile:

```
    def layout(self):
        if "CONAN_CMAKE_LAYOUT_BUILDDIR" in os.environ:
            cmake_layout(
                self, build_folder="build/{}".format(os.environ["CONAN_CMAKE_LAYOUT_BUILDDIR"]))
        else:
            cmake_layout(self)
```

With this hack, by setting the appropiate envvar, I can set the build folder to be whatever I want.

Would it be possible to expose a setting `tools.cmake.cmake_layout:build_folder` to set the cmake.build_folder property directly, without hacks?

Finally, when combining CMakeToolchain with cmake_layout, the name of the cmake preset is defined in the same way as the name of the build folder, so even with my hack, the name of the cmake doesn't match the build folder and again, makes it difficult to track what the preset is, or may have collisions with other preset names. As an example:

```
"configurePresets": [
        {
            "name": "linux-x86_64-gcc-10-debug",
            "displayName": "'linux-x86_64-gcc-10-debug' config",
            "description": "'linux-x86_64-gcc-10-debug' configure using 'Ninja' generator",
            "generator": "Ninja",
            "cacheVariables": {
                "CMAKE_BUILD_TYPE": "Debug"
            },
            "toolchainFile": "/workspace/build/asdf/linux-x86_64-gcc-10/Debug/generators/conan_toolchain.cmake",
            "binaryDir": "/workspace/build/asdf/linux-x86_64-gcc-10/Debug"
        }
    ],
```

Note that the toolchainfFile and the binaryDir point to `build/asdf` thanks to my hack, but the name of the preset doesn't have the `asdf` string. 

I link a discussion in slack about this
https://cpplang.slack.com/archives/C41CWV9HA/p1683276562027779

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @jjcasmar 

Thanks for your suggestion.

I have a quick question: why not using the ``--output-folder``? That allows to define any folder name as you wish. Internally you would have still the variability you want but that makes total sense, for example, inside a ``--output-folder=benchmarking`` you can still have some debug/release, static/shared variability for example.

**Comment by jjcasmar** [comment]
CMake preset wouldn't have the `output-folder` name in the preset name, so it will still have preset name conflicts, and difficult to parse the name in some cases

**Comment by memsharded** [comment]
Hi @jjcasmar 

Sorry this wasn't followed up. 
There is the ``conf`` ``tools.cmake.cmake_layout:build_folder_vars`` that you can set to include any ``settings`` or ``options``, so that will make different folders including them. It seems it would be good for your use case, can you please check and if so, close this ticket? Please let us know, thanks!

**Comment by Noxitu** [comment]
Hey, I also found myself wishing for this feature and for now Ive added such config on my recipe side.

For me the main reason for it was trying to get out of source build use exact path I want with `-of path -c tools.cmake.cmake_layout.build_folder_vars=[] -c build_folder=.` (somewhat hoping that there is no code that would do something undesired with the dot directory).

**Comment by memsharded** [comment]
Hi @jjcasmar and @Noxitu 

I think this was recently fixed in latest Conan 2.3, please check the new conf:

```
(conan2_310) λ conan config list | grep build_folder
tools.cmake.cmake_layout:build_folder: (Experimental) Allow configuring the base folder of the build for local builds
tools.cmake.cmake_layout:build_folder_vars: Settings and Options that will produce a different build folder and different CMake presets names
```

So the new ``tools.cmake.cmake_layout:build_folder`` can be now used as the full definer of the build-folder, even overriding the recipe ``cmake_layout()`` default "build" folder.

So I am closing this ticket as solved, please check the new ``tools.cmake.cmake_layout:build_folder`` conf, and if there are any issues, please open a new ticket referring to the new conf. Many thanks!
